### PR TITLE
Enabled customers to login with an valid domain using Service Principal

### PR DESCRIPTION
**Pull request by msJinLei**
## Description
Fix https://github.com/Azure/azure-powershell/issues/20728
### Current
image
### To be
image
### Current
image
#### v1
image
#### v2
image

## OpenID Config Debug Log
image

## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            this.Tenant = tenantDomain;
+            this._httpClientFactory = httpClientFactory;
+        }
+
+        public OpenIDConfiguration(string tenantDomain, IHttpOperationsFactory httpClientFactory = null) :this(DefaultAuthority, DefaultPath, tenantDomain, httpClientFactory)
+        {
+
+        }
+
+        public virtual async Task<string> Open(IHttpOperationsFactory httpClientFactory)
```
Comment (vidai-msft) [inline src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs:52]: Why this method is virtual?

File: src/Accounts/Accounts/Models/RMProfileClient.cs [hunk 2]
```diff
@@ -168,38 +172,44 @@ public bool TryRemoveContext(IAzureContext context)
...
+                            var tid = openIDConfigDoc.TenantId;
+                            token = new RawAccessToken()
+                            {
+                                AccessToken = token.AccessToken,
+                                LoginType = token.LoginType,
+                                TenantId = tid,
+                                UserId = token.UserId,
+                                HomeAccountId = token.HomeAccountId,
+                            };
+                            tenantIdOrName = token.TenantId;
```
Comment (vidai-msft) [inline src/Accounts/Accounts/Models/RMProfileClient.cs:197]: The tenantIdOrName looks like no need to be set again.

Comment (msJinLei) [reply src/Accounts/Accounts/Models/RMProfileClient.cs:197]: tenantIdOrName is used after this setting.

File: src/Accounts/Accounts/Models/RMProfileClient.cs [hunk 3]
```diff
@@ -168,38 +171,44 @@ public bool TryRemoveContext(IAzureContext context)
...
+                    IAccessToken token = null;
+                    try
                     {
-                        var tenants = ListAccountTenants(account, environment, password, promptBehavior, promptAction);
-                        var matchesName = tenants.Where(t => t.GetPropertyAsArray(AzureTenant.Property.Domains)
-                                                                                           .Contains(tenantIdOrName, StringComparer.InvariantCultureIgnoreCase));
-                        var homeTenants = matchesName.FirstOrDefault(t => t.IsHome);
-                        var tenant = homeTenants ?? matchesName.FirstOrDefault();
-                        if (tenant == null || tenant.Id == null)
+                        token = AcquireAccessToken(
```
Comment (isra-fel) [inline src/Accounts/Accounts/Models/RMProfileClient.cs:177]: If user input a domain name, will AcquireAccessToken() still be successful?

File: src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+using Newtonsoft.Json.Linq;
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Commands.Profile.Utilities
+{
+    internal class OpenIDConfiguration : IOpenIDConfiguration
+    {
+        private const string DefaultAuthority = "https://login.microsoftonline.com/";
```
Comment (isra-fel) [inline src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs:28]: Does this work for regional clouds? Like China cloud.

File: src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            Uri url = new Uri(Authority, Path);
+            if (!Uri.IsWellFormedUriString(url.AbsoluteUri, UriKind.Absolute))
+            {
+                throw new UriFormatException(string.Format(Resources.InvalidOpenIDConfigUri, url));
+            }
+            string content = await httpClientFactory.ReadAsStringAsync(url);
+            var jObject = JsonConvert.DeserializeObject<JObject>(content);
+            foreach (JProperty property in jObject.Properties())
+            {
+                if (string.Compare(property.Name, "issuer") == 0)
```
Comment (isra-fel) [inline src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs:63]: Use string.Equals instead

File: src/Accounts/Accounts/Utilities/IOpenIDConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+using Microsoft.Azure.Commands.Common.Authentication;
+
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Commands.Profile.Utilities
+{
+    public interface IOpenIDConfiguration
```
Comment (isra-fel) [inline src/Accounts/Accounts/Utilities/IOpenIDConfiguration.cs:20]: Would be great to add documents

File: src/Accounts/Accounts/Properties/Resources.resx [hunk 7]
```diff
@@ -574,4 +574,13 @@
   <data name="KeyStoreLoadingError" xml:space="preserve">
     <value>KeyStore cannot be loaded from storage. Please check the keystore file integrity or system compablity. The functions relate to context autosaving may be affected.</value>
   </data>
+  <data name="InvalidOpenIDConfigUri" xml:space="preserve">
+    <value>The uri is invalid: {0}</value>
+  </data>
+  <data name="OpenIDConfigResponseError" xml:space="preserve">
+    <value>Failed to find the 'issuer‘ from the OpenID configuration response: {0}.</value>
```
Comment (isra-fel) [inline src/Accounts/Accounts/Properties/Resources.resx:581]:
```suggestion
    <value>Failed to query tenant information by domain name: unable to find the 'issuer' from the OpenID configuration response: {0}.</value>
```

File: src/Accounts/Accounts/Properties/Resources.resx [hunk 8]
```diff
@@ -574,4 +574,13 @@
...
+  </data>
+  <data name="TenantIdNotFoundinIssuer" xml:space="preserve">
+    <value>Fail to parse tenant Id from issuer {0}.</value>
```
Comment (isra-fel) [inline src/Accounts/Accounts/Properties/Resources.resx:584]:
```suggestion
    <value>Failed to parse tenant ID from issuer {0}.</value>
```

File: src/Accounts/Accounts/Models/RMProfileClient.cs [hunk 9]
```diff
@@ -168,38 +171,44 @@ public bool TryRemoveContext(IAzureContext context)
...
+                            tenantIdOrName,
+                            password,
+                            promptBehavior,
+                            promptAction);
+
+                        if (!Guid.TryParse(tenantIdOrName, out Guid _))
                         {
-                            string baseMessage = string.Format(ProfileMessages.TenantDomainNotFound, tenantIdOrName);
-                            var typeMessageMap = new Dictionary<string, string>
+                            var tid = openIDConfigDoc.TenantId;
```
Comment (isra-fel) [inline src/Accounts/Accounts/Models/RMProfileClient.cs:187]: Can we add some debug logs?

File: src/Accounts/Accounts/Models/RMProfileClient.cs (hunk 3 above)
Comment (msJinLei) [reply src/Accounts/Accounts/Models/RMProfileClient.cs:177]: > If user input a domain name, will AcquireAccessToken() still be successful?

yes

File: src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs (hunk 4 above)
Comment (msJinLei) [reply src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs:28]: > Does this work for regional clouds? Like China cloud.

Will use ActiveDirectoryAuthority as CLI does

**Comment by isra-fel** [comment]
In the "To be" scenario of case 2, it is also worth letting the user double check if the tenant domain is correctly typed.

**Comment by msJinLei** [comment]
> In the "To be" scenario of case 2, it is also worth letting the user double check if the tenant domain is correctly typed.

@isra-fel What about to change the error message of case2 to the following statement?
image

File: src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+        public string TenantId
+        {
+            get
+            {
+                if (tenantId == null)
+                {
+                    if (_httpClientFactory != null)
+                    {
+                        var issuer = Open(_httpClientFactory).Result;
```
Comment (isra-fel) [inline src/Accounts/Accounts/Utilities/OpenIDConfiguration.cs:77]:
```suggestion
                        var issuer = Open(_httpClientFactory).ConfigureAwait(false).GetAwaiter().GetResult();
```

**Review by isra-fel — CHANGES_REQUESTED** [review]

**Comment by msJinLei** [comment]
Thanks @isra-fel, has address review comments.

**Review by isra-fel — DISMISSED** [review]

**Review by isra-fel — APPROVED** [review]

**Pull request merged**